python/mxnet/rnn/rnn_cell.py

```diff
@@ -0,0 +1,172 @@
+        if name not in self._params:
+            self._params[name] = symbol.Variable(name, **kwargs)
+        return self._params[name]
+
+
+class BaseRNNCell(object):
+    """Abstract base class for RNN cells"""
+    def __call__(self, inputs, states, params, prefix=''):
+        """construct symbol"""
+        raise NotImplementedError()
+
+    @property
+    def state_shape(self):
+        """shape(s) of states"""
+        raise NotImplementedError()
+
+    @property
+    def output_shape(self):
+        """shape(s) of output"""
+        raise NotImplementedError()
+
+    def begin_state(self, prefix='', init_sym=symbol.zeros, **kwargs):
+        """initial state"""
+        state_shape = self.state_shape
+        def recursive(shape, c):
+            if shape is None:
+                c[0] += 1
+                return init_sym(name='%sbegin_state_%d'%(prefix, c[0]), **kwargs)
+            elif isinstance(shape, tuple):
+                assert len(shape) == 0 or isinstance(shape[0], numeric_types)
+                c[0] += 1
+                return init_sym(name='%sbegin_state_%d'%(prefix, c[0]), shape=shape, **kwargs)
+            else:
+                assert isinstance(shape, list)
+                return [recursive(i, c) for i in shape]
+
+        return recursive(state_shape, [-1])
+
+    def _get_activation(self, x, activation, **kwargs):
+        if isinstance(activation, string_types):
+            return symbol.Activation(x, act_type=activation, **kwargs)
+        else:
+            return activation(x, **kwargs)
+
+class RNNCell(BaseRNNCell):
+    """Simple recurrent neural network cell"""
+    def __init__(self, num_hidden, activation='tanh'):
+        self._num_hidden = num_hidden
+        self._activation = activation
+        self._counter = 0
+
+    @property
+    def state_shape(self):
+        return (0, self._num_hidden)
+
+    @property
+    def output_shape(self):
+        return (0, self._num_hidden)
+
+    def __call__(self, inputs, states, params, prefix=''):
+        W = params.get('%si2h_weight'%prefix)
+        B = params.get('%si2h_bias'%prefix)
+        U = params.get('%sh2h_weight'%prefix)
+        name = '%st%d_'%(prefix, self._counter)
+        i2h = symbol.FullyConnected(data=inputs, weight=W, bias=B,
+                                    num_hidden=self._num_hidden,
+                                    name='%si2h'%name)
+        h2h = symbol.FullyConnected(data=states, weight=U, no_bias=True,
+                                    num_hidden=self._num_hidden,
+                                    name='%sh2h'%name)
+        output = self._get_activation(i2h + h2h, self._activation,
+                                      name='%sout'%name)
+
+        self._counter += 1
+        return output, output
+
+class LSTMCell(BaseRNNCell):
+    """LSTM cell"""
+    def __init__(self, num_hidden):
+        self._num_hidden = num_hidden
+        self._counter = 0
+
+    @property
+    def state_shape(self):
```

Comment:
Suggested comment:
"""LSTM has two internal states, typically called "cell" and "hidden".  Here we're setting them both to the same size."""